File: old/binary_tree/count_nodes/count_nodes.cpp
#include <iostream>

using namespace std;

class BinaryTree {
public: 
    int val;
    BinaryTree* left = nullptr;
    BinaryTree* right = nullptr;

    BinaryTree(int value, BinaryTree* l = nullptr, BinaryTree* r = nullptr) {
        val = value;
        left = l;
        right = r;
    }

    void printTreeInOrder() {
        if (left != nullptr) {
            left->printTreeInOrder();
        } 

        cout << val << " ";

        if (right != nullptr) {
            right->printTreeInOrder();
        }
    }
};

int countNodes(BinaryTree* root) {
    if (root == nullptr) {
        return 0;
    } else {
        int leftCount = countNodes(root->left);
        int rightCount = countNodes(root->right);
        return 1 + leftCount + rightCount;
    }
}

int main() {

    BinaryTree* root = new BinaryTree(
        5,
        new BinaryTree(3, new BinaryTree(1), new BinaryTree(4)),
        new BinaryTree(7, new BinaryTree(6), new BinaryTree(8))
    );



    if (root != nullptr) {
        root->printTreeInOrder();
    }
    cout << endl;

    cout << "total nodes: " << countNodes(root);

    return 0;
}

File: old/binary_tree/count_nodes/count_nodes.py
class BinaryTree:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def count_nodes(root):
    if root is None:
        return 0
    else:
        left_node_count = count_nodes(root.left)
        right_node_count = count_nodes(root.right)
        return 1 + left_node_count + right_node_count


def main():
    # Tree: 5 -> 4 -> 3 -> 2 -> 1
    # Tree:
    #        5
    #      /   \
    #     2     8
    #    / \   / \
    #   1   4 6   12
    #      /
    #     3
    root = BinaryTree(
        5,
        BinaryTree(2, BinaryTree(1), BinaryTree(4, BinaryTree(3))),
        BinaryTree(8, BinaryTree(6), BinaryTree(12)),
    )

    print(f"total nodes: {count_nodes(root)}")


main()
